text/0061-functions.md

```diff
@@ -0,0 +1,295 @@
+Policies that use a large number of functions may be hard to read.
+
+## Alternatives
+
+### Type annotations
+We could require/allow Cedar functions to have type annotations, taking type definitions from either the schema or allowing them to be inline.
+
+Example:
+
+Schema:
+```
+type SemVer = { major : Long, minor : Long, patch : Long };
+```
+Policy:
+```
+function semver(major : Long, minor : Long, patch : Long) -> Semver {
+  { major : major, minor : minor, patch : patch }
+};
+```
+
+This would allow functions to typechecked in absence of a use-site, and allow for easier user specification of intent.
+It would also probably result in clearer type checker error messages.
+
+This introduces the following questions:
+
+1. Do we allow `type` declaration in policies, or just in schemas?
+2. Are type annotations enforced dynamically à la Contracts, or are they just ignored at runtime?
+  1. If they are dynamically enforced, that implies access to the schema to unfold type definitions. It also may introduce redundant type checking.
+3. Are type annotations required or optional?
+4. Will we have types to support generics, i.e., polymorphism?
+
+
+Leaving them out for this RFC only precludes only one future design decision, making type annotations required.
+This decision feels unlikely, as we want Cedar to be useful without using a schema/the validator.
+Adding _optional_ type annotations in the future is backwards compatible, but mandating type annotations may not be, due to the use of polymorphic functions.
+
+### Let functions call other functions
+As long as cycles are forbidden and functions as arguments are disallowed, we could allow functions to call other functions without sacrificing termination.
+However, the potential complexity explosion is high, and it's backwards compatible to add this later.
+
+<!-- ### Make parse errors runtime errors
+Make (any subset of) the following errors runtime errors instead of parse errors: 
+
+1. Application of non-function
+2. Non-application of function
+3. Function application with incorrect arity -->
+
+### Naming
+Should these really be called `function`s? They are actually `macro`s.
+
+### Different namespaces for functions and variables
+In the style of a Lisp-2, we could have different namespaces for functions are variables.
+Assuming for the minute that functions-calling-functions is allowed:
+```
+function foo() {
+  ...
+};
+
+function bar(foo) {
+  foo(1) // No longer an error. `foo` is looked up in the function namespace not the variable namespace
+};
+
+function bar(foo) {
+  foo(foo) // Also fine
+}
+
+```
+We argue against this on principle of least surprise; very few modern languages work like this.
+
+## Unresolved Questions
+
+1. Do any functions ship with Cedar? If so, which ones? Leaving that out of this RFC and proposing it's handled equivalently to [RFC 58](https://github.com/cedar-policy/rfcs/pull/58)
+2. Can you import functions? Leaving that for a future RFC.
```

Comment:
Agree. Questions of management of functions, functions in namespaces, etc. are separable, just as management of type definitions from schemas for a standard library are separable.